### ISSUE TITLE
python mulit_digit_reader.py 12517.png error

### ISSUE THREAD

**Issue by bambrozio** [bdiesel/tensorflow-svhn#3]
Hi,
I've got one image from the downloaded dataset, and I'm trying to test it, but I'm facing the following error: 

Image used: `data/svhn/full/test/12517.png`
```
InvalidArgumentError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a mismatch between the current graph and the graph from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Assign requires shapes of both tensors to match. lhs shape= [160] rhs shape= [128]
	 [[node save/Assign_8 (defined at /Users/bambrozi/Downloads/tensorflow-svhn/multi_digit_reader.py:49) ]]
```

Do I have to crop the image or so? 
Can you please provide instructions on how to prepare the image before using it in the model?

Thanks!

**Comment by bambrozio** [comment]
Looks like it's because I tried to use your trained checkpoint you provided with my TensorFlow version `1.13.0`... I'm trying to retrain the module. It does work, I will post here and close this issue out.

Btw - I've had to update some parts of the code, in order to work with this TensorFlow version. As soon as I get it validated and working, I will send you a PR, thus you can merge if you want to get it updated.

Thanks for your effort. It's helping me a lot to understand CNN :)

**Comment by bambrozio** [comment]
2 Problems realised after training both train_classifier.py and train_regressor.py:

1 - After training the models (NUM_EPOCHS = 10):
```
  File "/Users/bambrozi/workspace/github.com/bambrozio/tensorflow-svhn/multi_digit_reader.py", line 77, in <module>
    raise IOError("Cannot find checkpoint file. Please run train_regressor.py")
IOError: Cannot find checkpoint file. Please run train_regressor.py
```

These are the files generated after training them:
```
ls -lh *ckp*
-rw-r--r--  1 bambrozi  staff   220B 13 May 00:07 checkpoint
-rw-r--r--  1 bambrozi  staff   4.3M 12 May 22:48 classifier.ckpt.data-00000-of-00001
-rw-r--r--  1 bambrozi  staff   1.1K 12 May 22:48 classifier.ckpt.index
-rw-r--r--  1 bambrozi  staff   113K 12 May 22:48 classifier.ckpt.meta
-rw-r--r--  1 bambrozi  staff   6.2M 13 May 00:07 regression.ckpt.data-00000-of-00001
-rw-r--r--  1 bambrozi  staff   1.4K 13 May 00:07 regression.ckpt.index
-rw-r--r--  1 bambrozi  staff   159K 13 May 00:07 regression.ckpt.meta

tree logs/
logs/
├── board_reg_train_writer
│   ├── events.out.tfevents.1557700325.Brunos-MBP
│   └── events.out.tfevents.1557700399.Brunos-MBP
├── board_reg_valid_writer
│   ├── events.out.tfevents.1557700325.Brunos-MBP
│   └── events.out.tfevents.1557700399.Brunos-MBP
├── board_train_writer
│   ├── events.out.tfevents.1557696950.Brunos-MBP
│   ├── events.out.tfevents.1557700325.Brunos-MBP
│   └── events.out.tfevents.1557700399.Brunos-MBP
├── board_valid_writer
│   ├── events.out.tfevents.1557696950.Brunos-MBP
│   ├── events.out.tfevents.1557700325.Brunos-MBP
│   └── events.out.tfevents.1557700399.Brunos-MBP
├── svhn_classifier_logs
│   ├── checkpoint
│   ├── train-0.data-00000-of-00001
│   ├── train-0.index
│   ├── train-0.meta
│   ├── validation-0.data-00000-of-00001
│   ├── validation-0.index
│   └── validation-0.meta
└── svhn_regression_logs
    ├── checkpoint
    ├── train-0.data-00000-of-00001
    ├── train-0.index
    ├── train-0.meta
    ├── validation-0.data-00000-of-00001
    ├── validation-0.index
    └── validation-0.meta
```

2 - When I change the variable `WEIGHTS_FILE = "regression.ckpt"` at multi_digit_reader.py to be `WEIGHTS_FILE = "regression.ckpt.data-00000-of-00001"` in a attempt to fix issue above (1), I got the error in the image:
```
`ValueError: Cannot feed value of shape (1, 64, 64, 4) for Tensor u'Placeholder:0', which has shape '(1, 64, 64, 3)'`
```
I can variate the entry image, and it will always complain about the shape. Do I have to preprocess my image somehow before using your model?
Can you sare a script that would put it eligible to be predicted? Also, why 64x64 and not 32x32?

**Comment by bdiesel** [comment]
Hey @bambrozio,  It's been a while since I've looked at this project.  If you have a PR for the tensorflow update please create a PR and I'll take a look at it there.  

It sounds like maybe the image you are using possibly has 4 channels instead of 3. Can you attach the image the image to your PR?

Thanks!

**Comment by bambrozio** [comment]
Sure, I will do @bdiesel.
Do you have any python snippet to convert images to this standard? `(1, 64, 64, 3)`
It would help me a lot at this stage... 
Thanks!

**Comment by bdiesel** [comment]
I don't have one on hand but I think you could use something like python openCV to do this.  https://pypi.org/project/opencv-python/  How did you get the 4 channel image to begin with?

**Comment by bdiesel** [comment]
There is also a reshape function in tensorflow,  you could do something like 
```resize_image = tf.reshape(image, [1, 64, 64, 3])```
to test the feed.

**Comment by bambrozio** [comment]
First I was doing:
```
import PIL.Image as Image
import matplotlib.pyplot as plt
sample_img = Image.open('/Users/bambrozi/Downloads/tmp/12517.png') # 12517.png lable: 22
image_size = (64, 64)
from scipy.ndimage import imread
from scipy.misc import imresize

image_data = imread(sample_img.filename)
lx, ly, _ = image_data.shape
n = 3
crop_face = image_data[lx // n: - lx // n, ly // n: - ly // n]
img_crop = imresize(crop_face, image_size)
plt.imshow(img_crop)
plt.show()
plt.imsave("/Users/bambrozi/Downloads/tmp/cropped/12517.png", img_crop)
````

Now I tried your suggestion either on my raw image and also in the one outputted from the code above. No one worked: 
From the original image:
`ValueError: Cannot reshape a tensor with 23373 elements to shape [1,64,64,3] (12288 elements) for 'Reshape' (op: 'Reshape') with input shapes: [49,159,3], [4] and with input tensors computed as partial shapes: input[1] = [1,64,64,3].`

From the cropped one:
`ValueError: Cannot reshape a tensor with 16384 elements to shape [1,64,64,3] (12288 elements) for 'Reshape' (op: 'Reshape') with input shapes: [64,64,4], [4] and with input tensors computed as partial shapes: input[1] = [1,64,64,3].`

Ideally should work from the original one...
I think I need a preprocessing before `reshape`ing it...
I have college exams next week. I will be able to get back working on it the next week.

**Comment by bambrozio** [comment]
Hi!
Getting back to it. Just an update: 

I manage to get the resize solved by using:
```
    image = tf.image.decode_png(tf.read_file(img_path), channels=3)
    image = tf.image.resize_images(image, [64, 64])
    image = tf.image.convert_image_dtype(image, dtype=tf.float32)
    sample_img = tf.reshape(image, [1, 64, 64, 3])
    with tf.Session().as_default():
        sample_img = sample_img.eval()
```

I've also trained the full model as in the files, and got:
Classifier:
```
2019-05-23 03:45:31.245703: step 32900, loss = 0.05  learning rate = 0.000106  (8.3 examples/sec; 30.73 sec/batch)
Mini-Batch Accuracy: 1.00%
Model saved in file: logs/svhn_classifier_logs/ckpt/classifier.ckpt
Test Accuracy: 0.91797%
```

Regressor:
```
2019-05-23 14:05:21.610912: step 120200, loss = 0.02  learning rate = 0.00  (1.1 examples/sec; 27.926 sec/batch)
Minibatch accuracy2: 1.00
Test accuracy: 0.93
```

But the single image prediction never get right. Eg.:
```
python multi_digit_reader_orig.py  /Users/bambrozi/Downloads/tmp/12517.png
(...)
Best Prediction 11 made in 0.0283141136169
```

But the image is:
![12517](https://user-images.githubusercontent.com/5923573/58258879-69b5b700-7d6b-11e9-8a4f-edc2143f910a.png)

I've tried loads of other (simpler) images, and no success yet.

I guess the resize / reshape I did is damaging the images... I couldn't plot it yet to see the results after the resize.

I will keep working on it.

**Comment by bambrozio** [comment]
btw - do you know how to extract one of the images from the test .mat dataset and save as an image? 
I've tried, but no success. I wanted to use one of them, as perhaps it would work without my patches...

I was wondering how did you test those API's. Don't you have there somewhere an image that works with those multi and single readers I can use to compare?

**Comment by bdiesel** [comment]
Sounds like great progress!  I think you're right about the resizing, seems like its something to do with the resizing.  Its been a while since I worked on this but I'll try to take a look the long weekend.

**Comment by bdiesel** [comment]
Have you checked out tensor-board?  https://www.tensorflow.org/guide/summaries_and_tensorboard.  I think was using this to inspect the images.

**Comment by bdiesel** [comment]
Also looks like the images are being rendered using PIL in single_digit_reader.py.  Maybe this would help you

**Comment by bambrozio** [comment]
Sure, I will.
Thanks!
Here's my branch in case you have some minutes to take a look, and perhaps suggest what can fix it: https://github.com/bambrozio/tensorflow-svhn/blob/dl-ca/multi_digit_reader.py#L27

Realise I'm also using `norm_pix` instead of `exp` at line 66. It was giving me another dimension error with the `exp`: 
```
        #feed_dict = {X: exp}
        feed_dict = {X: norm_pix}
```
The error before this change was:
```
ValueError: Cannot feed value of shape (1, 1, 64, 64, 3) for Tensor u'Placeholder:0', which has shape '(1, 64, 64, 3)'
```

> note: Ignore all the loads of files added. I will be sure to clean them up before sending you the PR when was the time :)

**Comment by bambrozio** [comment]
Just realise there's a pdf report here lol
Very good! I can learn much more now by reading it. Thanks again! :)

Btw - would be great if you still have the source code you've used to crop the pics you took from your neighbourhood as you state at: 

_"After the training was completed I decided to further experiment with my model using my own real world
images. The images below were captured while I walked around my neighborhood. I tested on a variety
of digit-sequences captured in the real world not just house numbers. The numbers were then cropped
using approximately the same technique used to feed the training model."_

it's exactly what I need to get the readers fixed 😅

**Comment by bambrozio** [comment]
Ok, here's the conclusion: 

When the image it's not squared, I crop it:
```
width, height = sample_img.size  # Get dimensions
    left = (width - 64) / 2
    top = (height - 64) / 2
    right = (width + 64) / 2
    bottom = (height + 64) / 2
    sample_img = sample_img.crop((left, top, right, bottom))
```

When it's squared. I simply resize: 
```
    basewidth = 64
    sample_img = sample_img.resize((basewidth, basewidth), Image.ANTIALIAS)
```

Then everything works. I can plot the images as well 😄 

But, a thing that was bothering me: Pretty much no image I try can be recognized. Only when I get the number very well cropped, removing all noise from the image...

I was wondering: how can my model have about 93% accuracy and I cannot predict any image? Even those ones that were used in the validation process 🤔 

Then I was digging in the trained model to understand what was going on, and I realise that, although it does separate the images for the test (not using it the training), it's cropping them according to the spec of the dataset, centralizing each one correctly to the number.

So, although it correctly separates the test images from the train, the model has high bias. It doesn't generalize well. Probably if I run the training again, but without touching the full dataset test images, probably my accuracy would be much less.

In order to validate this theory, I saved about 20 images from the test dataset, by using a snippet from the preprocessing script, and I could successfully use them on my single-image tests. Predicting correctly 8 from 20. And I've even needed to crop or resize then, as the image was preprocessed (perhaps this is why your readers don't have a resize step. Because it assumes the images will come treated).

Cool! I learned a lot and I guess with this last verification, this issue was addressed. So, closing out. There's nothing to do that justifies a PR, excepting upgrading libs, but I think doesn't worth as it would remove the authenticity of your work (great one by the way :)

Thank you very much again! I learned a lot from it! 😄
